dial.go

```diff
@@ -74,7 +73,19 @@ func dial(ctx context.Context, urls string, opts *DialOptions, rand io.Reader) (
 	opts = &*opts
 	if opts.HTTPClient == nil {
 		opts.HTTPClient = http.DefaultClient
+	} else if opts.HTTPClient.Timeout > 0 {
+		var cancel context.CancelFunc
+
+		ctx, cancel = context.WithTimeout(ctx, opts.HTTPClient.Timeout)
+		defer cancel()
+
+		opts.HTTPClient = &http.Client{
```

Comment:
Let's do the following in case more fields are added to `*http.Client`:

```
newClient = *opts.HTTPClient
newClient.Timeout = 0
opts.HTTPClient = &newClient
```